File: flutter_project/lib/features/organizations/domain/repositories/organization_repository.dart
/// Negocio Flex - Organization Repository Contract (Flutter / Dart)

import '../entities/organization_entity.dart';
import '../entities/organization_member_entity.dart';

abstract class OrganizationRepository {
  Future<List<OrganizationEntity>> getUserOrganizations(String userId);
  Future<OrganizationEntity?> getOrganizationById(String id);
  Future<OrganizationEntity?> getOrganizationBySlug(String slug);
  Future<OrganizationEntity> createOrganization({
    required String name,
    required BusinessType businessType,
    String? slug,
    String? description,
    String? phone,
    String? primaryColor,
    required String creatorUserId,
  });
  Future<OrganizationEntity> updateOrganization({
    required String id,
    required Map<String, dynamic> updates,
    required String callerUserId,
  });
  Future<List<OrganizationMemberEntity>> getMembers(String organizationId);
  Future<void> changeMemberRole({
    required String organizationId,
    required String targetUserId,
    required OrganizationRole newRole,
    required String callerUserId,
  });
  Future<void> removeMember({
    required String organizationId,
    required String targetUserId,
    required String callerUserId,
  });
}

File: src/features/organizations/domain/repositories/organization_repository.ts
/**
 * Negocio Flex - Organization Repository Contract (Fase 4)
 */

import {
  OrganizationEntity,
  OrganizationMemberEntity,
  OrganizationSettingsEntity,
  BusinessType,
  OrganizationRole,
} from '../entities/organization_entity';

export interface CreateOrganizationParams {
  name: string;
  businessType: BusinessType;
  slug?: string;
  description?: string;
  phone?: string;
  primaryColor?: string;
}

export interface IOrganizationRepository {
  /**
   * Obtiene todas las organizaciones donde el usuario es miembro activo
   */
  getUserOrganizations(userId: string): Promise<OrganizationEntity[]>;

  /**
   * Obtiene la información detallada de una organización por su ID
   */
  getOrganizationById(id: string): Promise<OrganizationEntity | null>;

  /**
   * Busca una organización por su slug público único
   */
  getOrganizationBySlug(slug: string): Promise<OrganizationEntity | null>;

  /**
   * Crea una organización de forma atómica (crea org + asigna owner + crea settings)
   */
  createOrganization(params: CreateOrganizationParams, creatorUserId: string): Promise<OrganizationEntity>;

  /**
   * Actualiza la información básica del negocio (requiere rol OWNER o ADMIN)
   */
  updateOrganization(id: string, updates: Partial<OrganizationEntity>, callerUserId: string): Promise<OrganizationEntity>;

  /**
   * Lista los miembros de la organización
   */
  getMembers(organizationId: string): Promise<OrganizationMemberEntity[]>;

  /**
   * Modifica el rol de un miembro (requiere rol OWNER y no permite dejar a la org sin owner)
   */
  changeMemberRole(
    organizationId: string,
    targetUserId: string,
    newRole: OrganizationRole,
    callerUserId: string
  ): Promise<void>;

  /**
   * Elimina a un miembro de la organización
   */
  removeMember(organizationId: string, targetUserId: string, callerUserId: string): Promise<void>;

  /**
   * Obtiene la configuración de la organización
   */
  getSettings(organizationId: string): Promise<OrganizationSettingsEntity | null>;

  /**
   * Actualiza la configuración de la organización
   */
  updateSettings(
    organizationId: string,
    settings: Partial<OrganizationSettingsEntity>,
    callerUserId: string
  ): Promise<OrganizationSettingsEntity>;
}
